### PR TITLE
team10 PM1.3

### PR DESCRIPTION
**Pull request by kostapsimoulis**

### CODE REVIEW

**Comment by smokhov** [comment]
Will open with this comment initial review of this PR and required mods. I will mark with the comments when done with "REVIEW DONE" in the end after a complete pass, FYI.

**Comment by smokhov** [comment]
(A) `.jar`s should be relocated to `lib/java/` with `lib/` being parallel with `src/`,

**Comment by smokhov** [comment]
(B), your "src" directory as far as IDE concerned and package structure is not `src/`, but rather `src/api/java/`. Thus Java packages should begin with `openiss.` as root, not `api.java.openiss.`.

**Comment by smokhov** [comment]
(C) Where does the `image_example.jpg` come from? What is its copyright status? Also it doesn't bear the required text in your spec. If the copyright allows, you'd need to alter the image to add the text and some documentation somewhere should point to the source of the image. Else it needs to be removed.

**Comment by smokhov** [comment]
(D) please rename `src/api/javascript` to a more common `src/api/js`

**Comment by smokhov** [comment]
(E) how and where does one invoke `docker-compose.yml`? Should it be in the `java/` directoy? Or more suitable for `tools/ci/java` or `tools/java`?

**Comment by smokhov** [comment]
(F) what is `etc/supervisor/conf/d/soap.conf`?

**Comment by smokhov** [comment]
(G) where is the build system? I require to have whatever project build files you are using, be it an IDE, Makefiles, or whatever. I need the Eclipse, IntelliJ, or OpenESB project files, whatever you've been using, to build and compile, in the repo.

**Comment by smokhov** [comment]
(H) commits like 23d859e followed by 1c9f39e are generally useless and should be purged from the history. They don't bring any value of something added then deleted and the net result no files in the final PR, but with still polluted and confusing history since none of those files got actually used anywhere. In the final PR, such commits should be `squash`ed and `force-push`ed. You don't have to do it, I can do myself when working on your final PR.

File: src/api/java/openiss/ws/soap/service/OpenISSSOAPService.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+@WebService
+@BindingType(value = "http://java.sun.com/xml/ns/jaxws/2003/05/soap/bindings/HTTP/")
+@SOAPBinding(style = SOAPBinding.Style.RPC)
+public interface OpenISSSOAPService {
+
+    @WebMethod
+    byte[] getFrame(@WebParam(name = "type") String type);
+
+    @WebMethod
+    String getFileName(@WebParam(name = "type") String type);
```
Comment (smokhov) [inline src/api/java/openiss/ws/soap/service/OpenISSSOAPService.java]: (I) `getFileName()` WebMethod was not in the spec; why do you need this service call? At the same time, the required `mixFrame()` service method is not even defined in the interface and implemented as a stub. I understand you may not have had the time to deal with it, but its declaration should still be in the interface.

**Comment by smokhov** [comment]
(J) please produce (and update as necessary) an `openiss.wsdl` document corresponding to the SOAP service after refactored and merged with #45 ; place the document for now under `src/api/java/ws`

**Comment by smokhov** [comment]
Initial REVIEW DONE

**Comment by kostapsimoulis** [comment]
(G) where is the build system? I require to have whatever project build files you are using, be it an IDE, Makefiles, or whatever. I need the Eclipse, IntelliJ, or OpenESB project files, whatever you've been using, to build and compile, in the repo.

Totally agreed. The issue is that most of team members are not familiar how to setup a java project in different environments and if we commit the project files we get a lot of git conflicts. We currently use IntelliJ. Do you want us to add the project files in .idea or src/api/java/.idea ? Ideally we want to use maven but for now we can add a Makefile or README how to compile using command line.

(A) .jars should be relocated to lib/java/ with lib/ being parallel with src/,

src/api/lib/java or src/lib/java ? If we add in src/lib/java then it's outside of scope of the IntelliJ project

(E) how and where does one invoke docker-compose.yml? Should it be in the java/ directoy? Or more suitable for tools/ci/java or tools/java?

It can be added anywhere but if we add it in tools/ci/java or tools/java then it would have some relative paths such as ../../src/api/java/ and it would not be accessible from the IDE (Some IDEs detect docker and have integration). In any case we can move it in tools/java since it's currently using for development and not for CI.

(F) what is etc/supervisor/conf/d/soap.conf?

The original docker image is using supervisor in order to startup some programs that need to be running all the time. This file is mapped using docker-compose.yml and it basically adds the SOAP service inside docker container and it maps the port to localhost. So you can access http://localhost:9090/openiss?wsdl directly on your workstation while this java service is running inside the container.

(H) commits like 23d859e followed by 1c9f39e are generally useless and should be purged from the history. They don't bring any value of something added then deleted and the net result no files in the final PR, but with still polluted and confusing history since none of those files got actually used anywhere. In the final PR, such commits should be squashed and force-pushed. You don't have to do it, I can do myself when working on your final PR.

I agree that they are useless but reason we did that is because 3 of my team members spend all day working on SOAPServerClient that we will eventually scrap because the other version in paul_soap_example branch was easier to run and configure. They are concerned that if I didn't merge their branch, their commits wouldn't show on the history and it wouldn't be fair to them because some parts of their code were actually copied and pasted to the other branch.

(B), (C), (D) Agreed, we will work on those changes.

**Comment by smokhov** [comment]
> > (G) where is the build system? I require to have whatever project 
> > build files you are using, be it an IDE, Makefiles, or whatever. I 
> > need the Eclipse, IntelliJ, or OpenESB project files, whatever you've 
> > been using, to build and compile, in the repo.
>
> Totally agreed. The issue is that most of team members are not familiar 
> how to setup a java project in different environments and if we commit 
> the project files we get a lot of git conflicts. We currently use 
> IntelliJ. Do you want us to add the project files in .idea or 
> src/api/java/.idea ? Ideally we want to use maven but for now we can add 
> a Makefile or README how to compile using command line.

Does everybody use IntelliJ or different IDE? If all use the same IDE, 
there should be very little problem. Simply avoid committing user-specific 
settings and UI preferences, but more general generic project 
configurations.

A simple Makefile is preferred for non-IDE builds, such as automated CI 
builds. Maven/gradle or ant would also work.

> > (A) .jars should be relocated to lib/java/ with lib/ being parallel 
> > with src/,
>
> src/api/lib/java or src/lib/java ? If we add in src/lib/java then it's 
> outside of scope of the IntelliJ project

This would be src/api/java parallel to lib/java both having the same root 
directory. Can't IntelliJ be told to where look for .jars, such as 
external .jars, etc.? Eclipse can.

> > (E) how and where does one invoke docker-compose.yml? Should it be in 
> > the java/ directoy? Or more suitable for tools/ci/java or tools/java?
>
> It can be added anywhere but if we add it in tools/ci/java or tools/java 
> then it would have some relative paths such as ../../src/api/java/ and 
> it would not be accessible from the IDE (Some IDEs detect docker and 
> have integration). In any case we can move it in tools/java since it's 
> currently using for development and not for CI.

I don't have a problem with using relative paths, it's just it's not a 
part of the the OpenISS Java API, so it should reside elsewhere. Again, I 
hope IntelliJ is intelligent enough to be look in different directories in 
the project settings?

Maybe a better place for it to be tools/docker/java/docker-compose.yml?

Perhaps you should show me which IDEs and how they integrate with docker.

> > (F) what is etc/supervisor/conf/d/soap.conf?
>
> The original docker image is using supervisor in order to startup some 
> programs that need to be running all the time. This file is mapped using 
> docker-compose.yml and it basically adds the SOAP service inside docker 
> container and it maps the port to localhost. So you can access 
> http://localhost:9090/openiss?wsdl directly on your workstation while 
> this java service is running inside the container.

So this related to (E). If so, they should move together, no?

> > (H) commits like 23d859e followed by 1c9f39e are generally useless and 
> > should be purged from the history. They don't bring any value of 
> > something added then deleted and the net result no files in the final 
> > PR, but with still polluted and confusing history since none of those 
> > files got actually used anywhere. In the final PR, such commits should 
> > be squashed and force-pushed. You don't have to do it, I can do myself 
> > when working on your final PR.
>
> I agree that they are useless but reason we did that is because 3 of my 
> team members spend all day working on SOAPServerClient that we will 
> eventually scrap because the other version in paul_soap_example branch 
> was easier to run and configure. They are concerned that if I didn't 
> merge their branch, their commits wouldn't show on the history and it 
> wouldn't be fair to them because some parts of their code were actually 
> copied and pasted to the other branch.

I see. These commits will still reside in the history of the respective 
branches in your fork, so they won't be lost from the history. (Though 
they should have been the ones doing the copy-pasting and commits to that 
other branch...). I'll decide on this later.

> (B), (C), (D) Agreed, we will work on those changes.

Don't forget (I) and (J)...

-- 
Serguei Mokhov, PhD, Manager, Networks & Security, AITS | /~\    The ASCII
Affiliate @ Computer Science and Software Engineering & | \ / Ribbon Campaign
Concordia Institute for Information Systems Engineering |  X    Against HTML
Concordia University, Montreal, Quebec, Canada          | / \      Email!
http://www.cs.concordia.ca/~mokhov
---
Co-founder, CCIFF, http://cciff.ca
ET Summit: http://cciff.ca/2017/festival/et/
https://www.youtube.com/channel/UCcJ-jSP6Cya8o-aWoB4l3Pw
---
Technical Lead
http://mdreams-stage.com
http://users.encs.concordia.ca/~gipsy
http://marf.sf.net | http://sf.net/projects/marf

**Pull request merged**